File: tlb/07/solutions/shortest_path.v1.pl
#!/usr/bin/perl -w
# find shortest path between two towns

die "Usage: $0 <start> <finish>\n" if @ARGV != 2;
$start = $ARGV[0];
$finish = $ARGV[1];

while (<STDIN>) {
    /(\S+)\s+(\S+)\s+(\d+)/ || next;
    $distance{$1}{$2} = $3;
    $distance{$2}{$1} = $3;
}

$shortest_journey{$start} = 0;
$route{$start} = "";
@unprocessed_towns = keys %distance;
$current_town = $start;
while ($current_town && $current_town ne $finish) {
    @unprocessed_towns = grep {$_ ne $current_town} @unprocessed_towns;
    
    foreach  $town (@unprocessed_towns) {
        if (defined $distance{$current_town}{$town}) {
        	my $d = $shortest_journey{$current_town} + $distance{$current_town}{$town};
        	if (!defined $shortest_journey{$town} || $shortest_journey{$town} > $d) {
        		$shortest_journey{$town} = $d;
        		$route{$town} = "$route{$current_town} $current_town";
        	}
        }
    }
    
    my $min_distance = 1e99;   # must be larger than any possible distance
    $current_town = "";
    foreach $town (@unprocessed_towns) {
        if (defined $shortest_journey{$town} && $shortest_journey{$town} < $min_distance) {
        	$min_distance = $shortest_journey{$town};
        	$current_town = $town;
        }
    }
}

if (!defined $shortest_journey{$finish}) {
    print "No route from $start to $finish.\n";
} else {
    print "Shortest route is length = $shortest_journey{$finish}:$route{$finish} $finish.\n";
}

File: tlb/08/solutions/shortest_path.v1.py
#!/usr/bin/python

import fileinput, re, sys, collections

(start,finish) = sys.argv[1:]
distance = {}
for line in sys.stdin:
    (town1,town2,dist) = line.split()
    distance[(town1,town2)] = int(dist)
    distance[(town2,town1)] = int(dist)

shortest_journey = {start:0}
route = {start:''}
unprocessed_towns = set(d[0] for d in distance.keys())
next_town = start
while next_town and next_town != finish:
    unprocessed_towns.remove(next_town)
    for town in unprocessed_towns:
        if (next_town,town) in distance:
            d = shortest_journey[next_town] + distance[(next_town,town)]
            if town not in shortest_journey or shortest_journey[town] > d:
                shortest_journey[town] = d
                route[town] = route[next_town] + " " + next_town
    next_town = min((shortest_journey[town],town) for town in unprocessed_towns if town in shortest_journey)[1]

if finish not in shortest_journey:
    print("No route from %s to %s" % (start, finish))
else:
    print("Shortest route is length = %s: %s %s." % (shortest_journey[finish], route[finish], finish))

